python/cudf/cudf/core/column/column.py

```diff
@@ -117,247 +264,57 @@ def _concat(cls, objs, dtype=None):
 
         head = objs[0]
         for obj in objs:
-            if not (obj.is_type_equivalent(head)):
+            if not (obj.dtype == head.dtype):
                 raise ValueError("All series must be of same type")
 
         # Handle strings separately
         if all(isinstance(o, StringColumn) for o in objs):
-            objs = [o._data for o in objs]
-            return StringColumn(
-                data=nvstrings.from_strings(*objs), name=head.name
-            )
+            objs = [o.nvstrings for o in objs]
+            return as_column(nvstrings.from_strings(*objs), name=head.name)
 
         # Filter out inputs that have 0 length
         objs = [o for o in objs if len(o) > 0]
-        nulls = sum(o.null_count for o in objs)
+        nulls = any(col.mask for col in objs)
         newsize = sum(map(len, objs))
-        mem = rmm.device_array(shape=newsize, dtype=head.data.dtype)
-        data = Buffer.from_empty(mem, size=newsize)
+
+        if is_categorical_dtype(head):
+            data = None
+            data_dtype = head.codes.dtype
+            children = (
+                as_column(
+                    rmm.device_array(shape=newsize, dtype=head.codes.dtype)
+                ),
+            )
+        else:
+            data_dtype = head.dtype
+            mem = rmm.device_array(shape=newsize, dtype=data_dtype)
+            data = Buffer(mem)
```

Comment:
Instead of doing this via `device_array` we should use the Buffer constructor directly and just calculate the proper size.
```suggestion
            mem = DeviceBuffer(size=newsize*data_dtype.itemsize)
            data = Buffer(mem)
```

Reply:
This looks done. Is that right @kkraus14?